docs/xss.rst

```diff
@@ -0,0 +1,95 @@
+-------------------------------------------------------------
+
+To trigger XSS vulnerability audits passively, set the ``VERIFY_XSS``
+environment variable to ``True``. Doing so triggers an XSS audit
+whenever a page object's ``q`` method is called. If errors are
+found on the page, an XSSExposureError is raised.
+
+.. note:: XSS vulnerabilities will only be found if you have injected the specific string
+XSS_INJECTION defined in bok_choy.page_object into your page content.
+
+You might already have some bok-choy tests written for your web application.
+Here is an example of a bok-choy test that will implicitly check for XSS vulnerabilities.
+It involves clicking on a button on page, and it is assumed that the user's name is
+displayed on the page with the button. If the user name is not properly escaped, the display
+of the name (which is data provided by the user and thus potentially malicious) can cause
+XSS issues.
+
+.. code-block:: python
+
+    from bok_choy.page_object import PageObject, XSS_INJECTION
+
+
+    class MyPage(PageObject):
+        def url(self):
+            return 'https://www.mysite.com/page'
+
+        def click_button(self):
+            """
+            Click on the button element (id="button").
+            On my example page this will trigger an ajax call
+            that updates the #output div with the text "yes!"
+            """
+            self.q(css='div#fixture button').first.click()
+            self.wait_for_ajax()
+
+        @property
+        def output(self):
+            """
+            Return the contents of the "#output" div on the page.
+            """
+            text_list = self.q(css='#output').text
+
+            if len(text_list) < 1:
+                return None
+            else:
+                return text_list[0]
+
+    class MyPageTest(WebAppTest):
+        def setUp(self):
+            """
+            Log in as a particular user.
+            """
+            super(MyPageTest, self).setUp()
+            self.user_name = XSS_INJECTION
+            self.username, self.user_id = self.log_in_as_user(self.user_name)
+
+        def test_button_click_output(self):
+            page = MyPage(self.browser)
+            page.visit()
+            page.click_button()
+
+            self.assertEqual(page.output, 'yes!')
+
+        def log_in_user(self, user_name):
+            """
+            Would be implemented to log in as a particular user
+            with a potentially malicious, user-provided name.
+            """
+            pass
+
+Before running your bok-choy tests, set the environment variable
+``VERIFY_XSS`` to ``True``.
+
+::
+
+    export VERIFY_XSS=True
+
+This will trigger an audit, looking for unescaped usage of XSS_INJECTION, whenever a call to
+``q`` is made. The audit will detect improper escaping both in HTML and JavaScipt embedded
+within HTML.
+
+In the case of the ``test_button_click_output`` test case in the example above,
```

Comment:
Possibly -- move the info in lines 90 -95 to BEFORE the test example, (changing the reference to "example below") and just end this document with the end of the code example.

"In the example below, an audit is done in the ... method calls, as each of those will call out to ``q``. The test case fails with an XSSExposureError if any XSS errors are found."